docs/developers/napari_architecture.md

```diff
@@ -0,0 +1,198 @@
+    without using `lambda event:`
+    """
+    # the code updating display code is quite complex and not relevant for this
+    # example thus has ommited.
+    print("Update number of dimensions displayed")
+
+# register our callback with the model
+dims.events.wind.connect(_update_display)
+
+# now, everytime dims.ndisplay is changed, _update_display is called
+dim.ndisplay = 3
+```
+
+This method is very customizable but requires a lot of boilerplate. The
+generic base model `EventedModel` was added to reduce this and
+"standardize" this change/emit pattern. The `EventedModel` provides the
+following features:
+
+* type validation and coersion on class instantiation and attribute assignment
+* event emission after successful attribute assignment
+
+Using `EventedModel` would reduce the above `Dim` class code to:
+
+```python
+class Dim(EventedModel):
+    """Dimensions object modeling slicing and displaying.
+
+    Parameters
+    ----------
+    ndim : int
+        Number of dimensions.
+    ndisplay : int
+        Number of displayed dimensions.
+    ...
+    """
+    ndim: float
+    ndisplay: float
+```
+
+Currently most of the models in `napari/components/` are `EventedModels` but
+not the layer models although there is intention to convert these to
+`EventedModels` in the future.
+
+## Qt classes
+
+Qt classes are responsible for all napari's user interface elements. There is
+generally one to one mapping between Python models and qt models in napari, for
+example Python model `Dims` and Qt model `QtDims`.
+The Qt class can register callbacks such that when an attribute of the
+corresponding Python model changes, the appropriate actions are taken.
+The Qt classes are also instantiated with a reference to
+the Python model, which gets updated directly when a field is changed via the
+GUI.
+
+For example, below is a code snippet showing the `QtDims` class instantiating
+with a reference to the python class `Dims` and registering the callback
+`_update_display`:
+
+```python
+class QtDims(QWidget):
+    """Qt view for the napari Dims model.
+
+    Parameters
+    ----------
+    dims : napari.components.dims.Dims
+        Dims object to be passed to Qt object.
+    ...
+
+    Attributes
+    ----------
+    dims : napari.components.dims.Dims
+        Dimensions object modeling slicing and displaying.
+    ...
+    """
+
+    def __init__(self, dims: Dims):
+
+        self.dims.events.ndisplay.connect(self._update_display)
+```
+
+## Vispy classes
+
+Vispy classes are responsible for drawing the canvas contents, thus need to be
+informed of any changes to Python models. They achieve this by connecting
+callbacks to Python model events just like Qt models.
+
+For example, below is a code snippet showing the `VispyCamera` class connecting
+the function `_on_ndisplay_change`:
+
+```python
+class VispyCamera:
+    """Vipsy camera for both 2D and 3D rendering.
+    """
+
+    def __init__(self):
```

Comment:
@andy-sweet silly question - do classes need to instantiate with `dims` to be able to connect a callback?